ui/raidboss/emulator/overrides/RaidEmulatorOverlayApiHook.js

```diff
@@ -33,11 +33,29 @@ export default class RaidEmulatorOverlayApiHook {
           // nextSignificantState is a bit inefficient but given that this isn't run every tick
           // we can afford to be a bit inefficient for readability's sake
           const combatantState = combatant.nextSignificantState(timestamp).toPluginState();
-          if (msg.ids && msg.ids.includes(parseInt(id, 16)))
-            combatants.push(combatantState);
-          else if (msg.names && msg.names.includes(tracker.combatants[id].name))
+          combatantState.ID = combatant.id;
+          combatantState.Name = combatant.name;
+          combatantState.Level = combatant.level;
+          combatantState.Job = combatant.job;
+          if ((msg.ids && msg.ids.length) || (msg.names && msg.names.length)) {
+            if (msg.ids && msg.ids.includes(parseInt(id, 16)))
+              combatants.push(combatantState);
+            else if (msg.names && msg.names.includes(tracker.combatants[id].name))
+              combatants.push(combatantState);
+          } else {
             combatants.push(combatantState);
+          }
```

Comment:
```suggestion
          if (!msg.ids || !msg.names)
            combatants.push(combatantState);
          else if (msg.ids && msg.ids.includes(parseInt(id, 16)))
            combatants.push(combatantState);
          else if (msg.names && msg.names.includes(tracker.combatants[id].name))
            combatants.push(combatantState);
```

I think this would be clearer like this.

It seems like there's an additional logic fix that you're making here too, in that previously we would have pushed a combatant if an id was specified but it wasn't included.  However, I'm not sure that the `msg.ids.length` additional change makes sense to me?  As an edge case, if I specify `ids: []` I think I'd expect that to return nothing rather return everything.

Reply:
This is meant to copy the logic of the entire getCombatants filter in both of the following parts of MiniParseEventSource.cs:
https://github.com/ngld/OverlayPlugin/blob/master/OverlayPlugin.Core/EventSources/MiniParseEventSource.cs#L117-L135
https://github.com/ngld/OverlayPlugin/blob/master/OverlayPlugin.Core/EventSources/MiniParseEventSource.cs#L263-L289

Unless I'm misreading that code (it's been 15 months since I wrote it, apparently), if an empty array is submitted it's treated the same as if the array is completely omitted?

Reply:
Ok!  Thanks for the reference.

I still think not nesting a conditional makes the logic easier to read, at least.

Reply:
I don't know how to get around nesting the conditionals without making the first conditional really long. It would basically have to be something like this?

```javascript
          if ((!msg.ids || !msg.ids.length) && (!msg.names || !msg.names.length)) {
            combatants.push(combatantState);
          else if (msg.ids && msg.ids.includes(parseInt(id, 16)))
            combatants.push(combatantState);
          else if (msg.names && msg.names.includes(tracker.combatants[id].name))
            combatants.push(combatantState);
```

Maybe instead move this up in scope so it's something like:

```javascript
      return new Promise((res) => {
        const combatants = [];
        // If no ids and names, or both are empty, return all combatants
        const allCombatants = (!msg.ids || !msg.ids.length) && (!msg.names || !msg.names.length);

        for (const id in tracker.combatants) {
...
          if (allCombatants) {
            combatants.push(combatantState);
          else if (msg.ids && msg.ids.includes(parseInt(id, 16)))
            combatants.push(combatantState);
          else if (msg.names && msg.names.includes(tracker.combatants[id].name))
            combatants.push(combatantState);
```

Reply:
What about:

```javascript
          const hasIds = msg.ids && msg.ids.length;
          const hasNames = msg.names && msg.names.length;
          if (!hasIds && !hasNames)
            combatants.push(combatantState);
          else if (hasIds && msg.ids.includes(parseInt(id, 16)))
            combatants.push(combatantState);
          else if (hasNames && msg.names.includes(tracker.combatants[id].name))
            combatants.push(combatantState);
```

Reply:
That also works, though I still think it makes sense to put those two `const` variables outside the for loop? No sense in checking them every iteration of the loop.

Reply:
Even better.  :+1:

Reply:
Hmm, this is functional but I'm not really a fan of the typing of the hasIds/hasNames vars?

![image](https://user-images.githubusercontent.com/6119598/120021071-9636e600-bfb8-11eb-83af-83e4e0faaabb.png)

Maybe make it more explicit?

```javascript
        const hasIds = msg.ids !== undefined && msg.ids.length > 0;
        const hasNames = msg.names !== undefined && msg.names.length > 0;
```

![image](https://user-images.githubusercontent.com/6119598/120021235-c9797500-bfb8-11eb-90ef-53ee503111b2.png)

Reply:
Sounds good to me.

Reply:
This has been updated.